[2단계 - 페이먼츠] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
안녕하세요 케빈!! 니야입니다 ~! 😄
주말은 잘 보내셨나요? 저는 평일 내내 잠을 거의 못자서 주말에 몰아서 숙면한 것 같아요 😃

주말 시작 쯤에 요청드려야 케빈이 조금 더 편하게 리뷰해주실 수 있을 텐데 주말 끝무렵에 제출하게 되어버렸네요..
기능 구현 내내 고민도 정말 많았고, 계속 무한 리팩토링 하느라 시간이 좀 걸린 것 같습니다 🤯

본 코드의 이해를 돕기 위해서 간단하게 컴포넌트 관계도를 간략하게 작성해봤어요! 세부 관계도는 `README`문서에 있습니다 :)

### ✔️ CardForm 페이지 (path: "/")

```
├── <CardPreview> (카드 프리뷰)
├── <CardFormScrollStyles> (하위 컴포넌트는 조건부 렌더링)
│   ├── <CardFormSection type="password"> (비밀번호)
│   ├── <CardFormSection type="cvcNumber"> (CVC 번호)
│   ├── <CardFormSection type="expirationPeriod"> (유효 기간)
│   ├── <CardFormSection type="cardCompany"> (카드사)
│   └── <CardFormSection type="cardNumber"> (카드 번호)
└── <Button type="submit"> (카드 등록 버튼: 조건부 렌더링 - isFormComplete일 때만)
```

### ✔️ RegisterComplete 페이지 (path: "/register-complete")

```
├── <CheckImgStyles> (체크 이미지)
├── <CompletionMessageStyles>
│   ├── <CompletionTitleStyles> (카드번호 첫 부분)
│   └── <CompletionTitleStyles> (카드사 이름)
└── <Button onClick={handleConfirm}> (초기 화면 이동 버튼)
```

## 🎯 페이먼츠

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계
- 재사용 가능한 Input Component를 개발한다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 테스트한다.
- 카드 정보를 효과적으로 렌더링 하기 위한 상태 관리를 경험한다.

### 2단계
- 다양한 Form 구성 요소들간의 상태를 효율적으로 관리한다.
- hooks API를 이용하여 상태 관리 로직을 구현한다.
- custom hooks를 생성하여 Form 관리 로직을 컴포넌트에서 분리하고 재사용한다.
- Controlled & Uncontrolled Components에 입각하여 Form을 핸들링한다.

---

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항(코드 컨벤션, 에러 핸들링 등)을 준수했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - [니야🐢의 페이먼츠 사이트](https://sooyeoniya.github.io/react-payments/)
  - [스토리북 문서 배포 사이트](https://68007e9dc28a16d0492e8575-pkvdmpvmjr.chromatic.com/)

### 리뷰 요청 & 논의하고 싶은 내용

#### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

고민과 해결 과정을 정리하다 보니 내용이 길어져, PR 본문에 모두 담기에는 PR을 읽기가 조금 불편하실 것 같았어요.
그래서 아예 docs 폴더에 별도로 정리해보았습니다! 아래 링크에서 확인 부탁드립니다! 🙇🏻‍♀️

- [CardInputGroup 컴포넌트 제거 및 각 컴포넌트 분리](https://github.com/sooyeoniya/react-payments/blob/step2/docs/troubleshooting%3A%20component%20separation.md)
- [useCardValidation 훅을 Context API로 리팩토링](https://github.com/sooyeoniya/react-payments/blob/step2/docs/troubleshooting%3A%20context-api.md)

#### 2) 이번 리뷰를 통해 논의하고 싶은 부분

우선, 기본적으로 학습 목표나 체크 포인트 기준으로 확인해주시면 감사할 것 같습니다!
아래는 제 코드에서 봐주셨으면 하는 부분에 대해 나열해봤어요 ㅎㅎ

**1️⃣ useEffect를 적절하게 사용했는지?**

- useEffect를 함부로 남용하지는 않았는지?
- 의존성 배열이 적절하게 값이 잘 들어갔는지?

**2️⃣ useState와 useRef를 적절하게 잘 사용했는지?**
 
- 제어와 비제어 컴포넌트를 적절히 사용했는지?
- 불필요한 상태를 가지고 있지는 않는지?

**3️⃣ 성능 최적화 관점에서 문제가 발생할만한 부분이 있는지?**

개발 과정에서 useCallback과 useMemo는 최대한 사용하지 않고 구현을 진행했습니다. 
물론, 함수 참조가 변경되지 않아야 하는 부분과 의존성 배열에 의해 함수 참조를 새로 생성해야 하는 경우에는 useCallback을 적용해주었습니다!

그 외에는 useCallback과 useMemo를 사용하지 않더라도 렌더링 측면에서 큰 문제가 발생하지 않았던 것 같아요.
전체 흐름을 테스트한 결과, LCP는 약 0.13ms, INP는 최대 32~40ms로 체감상 거의 딜레이 없이 동작했습니다.

하지만, Profiler에서 테스트한 결과, 모든 입력 폼 관련 컴포넌트가 입력할 때마다 매번 렌더링되고 있음을 확인했습니다. 
물론 이 부분은 입력 상태에 대한 유효성 검증과 입력 값의 추적을 위해 useState로 상태를 관리하는 과정에서 발생한 불가피한 부분이라고 생각합니다.

이를 해결하려면, 상태들을 서로 의존하지 않도록 분리해야 하지만, 지난 리뷰에서 케빈이 "지금 상태에서 하나의 Context로 유지해도 괜찮다"고 하셨고, 
현재 구조에서 Context를 여러 개로 분리하는 것은 과도한 분리로 보였습니다. 그래서 그대로 두기로 했습니다.

하지만, 이런 최적화 관점에서 제가 놓친 부분이 있을 수 있다고 생각합니다.
제 코드에서 성능 문제가 발생할 수 있는 부분이 있거나 더 최적화 할 수 있는 부분이 있다면, 확인해주시면 정말 감사하겠습니다 😃


**📌 페이지 단위의 Storybook UI 테스트에 대하여**

그리고 마지막으로, 케빈의 의견을 듣고 싶은 부분이 하나 있습니다! 😄

이번 미션에서는 컴포넌트에 대한 UI 테스트를 모두 Storybook으로 진행했습니다.
하지만, 이 과정에서 **페이지 단위의 컴포넌트도 UI 테스트를 하는 것이 맞는가?** 에 대한 의문이 들었습니다.

보통은 cypress와 같은 테스트 도구로 E2E 테스트를 많이 진행합니다.
이전 미션까지도 cypress를 통해 UI 단에서 사용자 흐름 테스트를 진행했기도 했구요!
이와 유사하게, Storybook 자체에서도 Interaction을 잘 활용하면 UI 단에서 흐름을 직접 테스트할 수 있습니다.

하지만, 이번 미션에서는 E2E 테스트에 대한 요구사항이 없었는데요. 그렇다보니 이런 고민을 생각하게 된 것 같아요.
왜냐하면, 현재 CardForm과 RegisterComplete 페이지 단위의 UI 테스트를 작성했는데요.
이 과정에서 Interaction이 따로 존재하지 않아서 초기 페이지만 보이기 때문에 의미가 없는 느낌이 들었어요.

제가 생각하는 Storybook의 주요 목적은 작은 단위의 컴포넌트를 분리하여 테스트하는 것에 있다고 생각합니다.
그래서 굳이 Interaction을 통해 페이지 전체 UI를 테스트해야 할지 고민이 됩니다. 
아예 페이지 단위의 컴포넌트 테스트는 지우는 것이 맞다는 생각도 들었어요. 
하지만, 한 편으로는 Storybook의 Interaction을 시도해보는 것도 좋은 의미가 있을 것 같다는 생각이 드네요.

그래도 현재 학습 목표에서 약간 벗어난 부분인 것 같기도 해서, 
다른 부분에 더 집중하는 것이 더 시간적으로 효율적일 것 같다는 생각이 들었어요. 
이게 시간이 얼마나 걸릴지 아직 감이 잘 오지 않는 부분이기도 해서 그런 것 같아요 🥲

**"페이지 단위의 컴포넌트 UI 테스트는 지운다"** vs **"Storybook의 Interaction을 통해 사용자 흐름 UI 테스트를 진행한다"**
이 의견에 대해 여러 조건을 고려해봤을 때 케빈의 생각은 어떠신지 궁금해요!

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도만이 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. Form 상태 관리 & Custom Hook 분리

- 반복되는 로직을 custom hook으로 분리했는가?
- hook 내부와 UI 컴포넌트의 **역할이 명확하게 분리**되어 있는가?
- 상태 흐름이 직관적이며, Form 전체를 일관되게 관리할 수 있는 구조인가?

### 2. 입력 UI 흐름과 UX

- 카드 번호 입력 시 필드 간 **자동 포커싱 이동**이 자연스럽게 동작하는가?
- 숫자만 입력 가능한 필드에서 제한, 에러 메시지, 유효성 피드백 등 **사용자 경험이 충분히 고려**되었는가?

### 3. 컴포넌트 구조 및 재사용성

- 컴포넌트가 명확한 **역할과 책임**을 가지며 과도하게 분리되거나 중첩되어 있지 않은가?
- Input, Button 등 재사용 가능한 컴포넌트가 잘 정의되어 있는가?

### 4. 상태 기반 유효성 검사 및 확인 버튼 활성화

- 모든 필드가 유효할 때만 확인 버튼이 **정확히 활성화/비활성화**되는가?
- 유효성 검사의 기준이 명확하고, **상태 변경에 따른 UI 반응**이 잘 연결되어 있는가?

wmakerjun committed:
docs: PR 템플릿 추가

sooyeoniya committed:
refactor: Input 컴포넌트에 key 값을 지정해 식별되도록 함

sooyeoniya committed:
refactor: 유니온 타입 활용해 Record 로 상수 타입 재선언

- interface로 길게 나열되어 있던 타입을 Record타입으로 간결하게 변경

sooyeoniya committed:
refactor: Input 컴포넌트 InputHTMLAttributes 타입 확장 적용

[장점]
- 확장성 및 재사용성 향상

[교차 타입(&) vs 인터페이스 확장(extends) 비교]
- 인터페이스를 확장하는 방식을 선택한 이유
: 타입 충돌했을 때 확장된 인터페이스 속성이 우선시 됨, 교차 타입의 경우 TS가 더 구체적인 타입을 선택하거나 오류 발생

sooyeoniya committed:
refactor: Input 에러 시 focus outline 제거하여 에러 border 잘보이게 수정

- 에러를 잘 인식하도록 하여 웹 접근성 및 UX 개선

sooyeoniya committed:
docs: README.md Step2 요구사항 업데이트

sooyeoniya committed:
fix: Input의 isError 속성 기본값 설정 및 옵셔널 제거

- 통일되지 않은 부분에 대해 일관성 유지하도록 함

sooyeoniya committed:
feat: Select 컴포넌트 UI 구현

- placeholder 특성상 옵셔널로 정의하되, 의미있는 기본값 제공하도록 함 -> undefined인 경우 '선택해주세요' 출력
- option 태그 스타일링 제한 (브라우저 자체에서 제한하고 있음)

sooyeoniya committed:
feat: Subtitle도 Description과 같이 null 값 명시적으로 허용

- 없는 경우가 존재하기 때문에 빈값('')을 넣는 것이 아닌 null로 처리하여 아예 요소가 렌더링 되지 않게 함 -> height 때문에 이 방식이 적합함

sooyeoniya committed:
refactor: CardInputGroup 컴포넌트 제거 및 각 컴포넌트 분리

[CardInputGroup 문제점]
- 다양한 타입에 대한 너무 많은 컴포넌트를 렌더링하고 있음
- 코드줄이 길어져서 가독성 저하
- 확장성을 고려했을 때 관심사 분리가 제대로 되지 않아 테스트도 어렵고 가독성이 매우 떨어짐, 유지보수 하기도 어려움

[CardInputGroup 제거 후 이점]
- 각각의 폼에 맞춰 분리가 되어 응집도 향상 및 확장성 용이해짐
- 불필요하게 CardInputGroup에서 모든 type, error, setXXXError를 받아 내부에서 분기 처리해주는 로직이 없어서 훨씬 깔끔
→ 이 책임을 모두 CardInputSection에게 위임함 (type은 그냥 CardInputSection 에서 한 번만 처리해주면 됨)
→ 각 컴포넌트와 관련된 error, setXXXError 상태만 넘기면 되므로 훨씬 Props 관리가 잘 됨

[앞으로 할 일]
- 각 컴포넌트에 error를 통채로 넘기는 것이 아니라 각 컴포넌트에 관련된 error 상태만 넘기도록 수정할 예정

sooyeoniya committed:
refactor: 컴포넌트 및 폴더 이름 변경

변경사항: Input -> Form
- 카드 섹션마다 Input만 있는 것이 아닌 Select도 있기 때문
- LMS 학습목표에도 'Form 구성 요소들'이라고 나와있기 때문에 통일해주는 것이 맞다고 판단함

sooyeoniya committed:
refactor: 유효성 검증 훅 관련 로직 네이밍 수정

변경사항: CardInputError -> CardValidation
- 유효성 검사하는 훅임을 좀 더 선언적으로 보여주도록 하기 위함
- error도 그냥 error가 아니라 validationErrors로 변경 (여러 개의 유효성 관련 에러 상태를 가지고 있다는 의미에서 validation 접두사 붙이고 복수형으로 변경)

sooyeoniya committed:
refactor: 각 컴포넌트에 맞는 error 상태만 가져오도록 수정

[추가 반영 사항]
관련 네이밍도 더 그 역할에 맞는 명칭으로 수정

[고민]
각각의 error 상태별로 훅을 모두 분리하는 것이 더 좋은 설계일까?

sooyeoniya committed:
refactor: 에러가 있는지 확인하는 로직 useCardValidation 내부에서 정의

[내부로 분리한 이유]
상태와 관련된 비즈니스 로직을 한 곳에 캡슐화하여 선언적으로 사용해주기 위함

[타입 안정성 고려]
undefined 또는 null인 경우 false로 변환해주기 위해 이중 부정 연산자(!!) 사용함

sooyeoniya committed:
chore: CardFormSection 이름 오타 수정

- 복수형 -> 단수형

sooyeoniya committed:
feat: CardPasswordInput UI 구현

sooyeoniya committed:
refactor: export default 위치 맞추기

sooyeoniya committed:
refactor: 상태 타입 네이밍 수정

- 접미사에 'State'를 붙여 상태를 위한 타입임을 좀 더 명확히 알리도록 함

sooyeoniya committed:
refactor: 카드 번호와 유효기간 각 Input 명칭 변경

- POSITION -> SEGMENT
segment는 분류된 그룹 하나 하나를 의미하므로 그 역할에 더 잘맞는 네이밍이라고 판단함

sooyeoniya committed:
docs: README.md 문서 오류 수정

sooyeoniya committed:
feat: isEmpty 유효성 로직 구현

- select 태그의 value가 비어있는 경우를 확인하기 위한 유효성 검증 로직

sooyeoniya committed:
feat: cardCompany, password에 대한 error 상태 get, set 구현

sooyeoniya committed:
feat: 카드사, 비밀번호 에러 상태 연동

sooyeoniya committed:
chore: Input storybook 에서 필요없는 속성 제거

- onChange는 UI 테스트에 영향을 주지 않는 요소임

sooyeoniya committed:
feat: 카드 번호, 비밀번호 상태 관리 로직 추가

- cardCompany의 경우 none '' 빈 값에 대해 명시적으로 추가해줘야 함
- 안그러면 as const 를 사용해서 타입 오류를 우회해줘야 하기 때문

sooyeoniya committed:
feat: companySelect에 대해 타입 가드 추가

- value값이 string에서 CardCompanyState로 타입이 좁혀져야 하기 때문에 타입 가드 생성
- as 타입 단언 쓰지 않고 하는 방법을 찾음
- 일단 비교할 값들을 배열로 저장해두고, 하나씩 순회하면서 하나라도 일치하는 경우에 true를 반환하도록 구현

sooyeoniya committed:
refactor: CardFormFields 폴더 내부 구조 변경

- 폴더 depth 줄이기

sooyeoniya committed:
feat: withCustomCardProvider에 기본 속성 추가

- cardCompany, password 상태 추가

sooyeoniya committed:
feat: CardFormFields story 구현

sooyeoniya committed:
design: 전체 height 및 scroll 설정

sooyeoniya committed:
design: 각 CardFormSection에 margin-bottom 추가

- 기존에 Title에 붙어있던 margin-top 을 없애고 CardFormSection의 margin-bottom으로 옮김
- 직접 UI 단에서 보면 Title에 붙어있을 경우 스크롤을 위로 끝까지 올렸을 때 위에 공간이 너무 많이 남아서 이상해보임
- 차라리 아래에 margin이 붙어 있는 것이 훨씬 이상적

sooyeoniya committed:
design: 전체 요소 크기 확대

sooyeoniya committed:
feat: 카드 회사에 따라 카드 색상 변경

- IC 칩 UI 구현
- 카드 회사에 따라 색상 변경 로직 구현
-> 단순 매핑이기 때문에 객체로 만들어서 매칭시켜줌

sooyeoniya committed:
feat: cardCompany 관련 Preview Story 추가

sooyeoniya committed:
refactor: cardCompany에 따른 색상 변경 책임 이동

- style 관련 객체 상수이므로 styled.ts 내부에서 직접 정의하여 사용함
-> 중복 제거 및 컴포넌트 내부 간결화

sooyeoniya committed:
design: 전체 padding 조정

sooyeoniya committed:
design: reset css 설정

sooyeoniya committed:
feat: 비밀번호 안보이게 수정

sooyeoniya committed:
docs: README.md 문서 수정

sooyeoniya committed:
refactor: 카드 타입 계산하는 로직 간결화

- every 메서드와 정규 표현식을 활용해 로직 수정

sooyeoniya committed:
refactor: CardValidationType에 대한 의존성 제거
현재 문제점: 여러 모듈에서 useCardValidation이라는 훅에 의존하여 타입을 받아오고 있음
해결책: 아예 상수파일로 분리하여 타입 가져오기

sooyeoniya committed:
fix: 카드 타입 식별 로직 오류 수정

- 완성되지 않은 상태에서 식별이 되는 오류가 발생하여 수정함

sooyeoniya committed:
refactor: useCardValidation 훅을 Context API로 리팩토링

[문제 상황]
- Preview에서도 에러 상태 필요
- 스토리북에서 에러 UI 테스트를 할 수 없음

[해결 방법]
- useCardValidation 훅을 커스텀 훅에서 Context API로 변경하기로 함

[장점]
- props drilling 줄이고 스토리북 테스트 용이성 확보
- CardFormSection 단위에서 유효성 검증 관리 방식 통일
- 기존 CardContext와 구조 일관성 맞춤

[추가 작업]
- 기존에는 하나의 useState로 객체 형태로 상태를 관리했으나, 각각의 상태를 개별 useState로 분리하여 독립적으로 관리되도록 수정

sooyeoniya committed:
refactor: Provider 위치 변경

- CardPaymentsPage에 있는 것이 모듈 단위로 상태가 제공된다는 의미에서 좀 더 나은 위치라고 판단

sooyeoniya committed:
feat: CardValidation에 대한 스토리북 커스텀 제공자 구현

- hasErrorByType은 오버라이딩 하여 구현
> defaultValue 내부에서 하면, 각 상태가 모두 false 상태이므로 그 값을 사용해 실행되기 때문에 에러가 발생하지 않음
> 밖에서 contextValue를 사용해 에러가 존재하는지 확인하는 로직을 실행해야 함. 그래야 에러 UI를 확인할 수 있음

- tsconfig.app.json의 include에 추가해줘야 하는이유
> 추가하지 않으면, React를 import 해야함
> 왜냐하면 jsx 변환 과정에서 React.createElement 호출로 변환을 하기 때문에 명시적으로 React가 필요하기 때문
> include에 작성해두면, React 17+ 부터 적용되는 새로운 JSX 변환이 자동으로 사용됨 (automatic runtime)

sooyeoniya committed:
fix: 카드 입력 에러일 때, 카드 타입이 뜨지 않도록 수정

- 입력값이 에러일 때도, 카드 번호만 16자면 preview에서 visa, master 구분이 뜨고 있음

sooyeoniya committed:
fix: storybook에서 CardValidation provider 적용하여 에러 수정

sooyeoniya committed:
feat: CardPreview 스토리북 상수 분리

- 카드 번호와 유효기간 번호 상수로 분리하여 통일성 주기
- 카드 색상 있을 때도 내부 글씨 보이도록 수정 -> 카카오뱅크 글씨 색상 다른 모습 보여주기 위함

sooyeoniya committed:
feat: 유효성 검증 시 여러 에러 메시지 받을 수 있도록 수정

- 객체 상태에서 각 입력폼 별로 errorMessage 추가

sooyeoniya committed:
feat: 바뀐 에러 상태 값에 맞춰 storybook 수정

sooyeoniya committed:
feat: 유효기간 유효성 검증 로직 추가

sooyeoniya committed:
feat: 모든 상태에 대한 유효성 검증 로직 추가

sooyeoniya committed:
feat: 유효 기간의 월(MM)이 2자리수가 아닌 경우 에러 처리

sooyeoniya committed:
feat: 입력 폼 onBlur 처리

sooyeoniya committed:
refactor: 입력 값 길이 검증 및 재사용 함수 분리

- 리팩토링 사항: PLACEHOLDER와 MAX_LENGTH 상수 공통으로 분리

sooyeoniya committed:
refactor: 상수와 타입 분리

sooyeoniya committed:
refactor: utils 파일 구조 리팩토링

sooyeoniya committed:
feat: Button 컴포넌트 구현

sooyeoniya committed:
feat: CardPaymentsPage에 Button 컴포넌트 적용

sooyeoniya committed:
refactor: 페이지 단위 컴포넌트 pages폴더로 분리 및 네이밍 변경

- CardPaymentsPage -> CardForm

sooyeoniya committed:
chore: 이미지 jpg -> svg 변경

- 더 경량화된 파일 확장자 적용

sooyeoniya committed:
feat: 등록 화면 구현 및 라우터 설정

sooyeoniya committed:
chore: CSS -> Styles 명칭 변경

- Styled-components 로 만든 컴포넌트 suffix를 CSS에서 Styles로 변경함

wmakerjun committed:
docs: PR 템플릿에 2단계 학습 목표 및 리뷰어 가이드 추가

sooyeoniya committed:
feat: 확인 버튼 클릭 시 카드 등록 로직 구현

sooyeoniya committed:
refactor: useCardType 제거 및 함수로 분리

- 불필요한 상태 관리 제거
- identifyCardType 함수 자체도 순수함수이므로 내부에서 바로 접근해도 괜찮음
- cardNumbers가 변경될 때만 함수 결과가 달라지므로 의존성이 명확

sooyeoniya committed:
feat: 확인 버튼 클릭 시 메인 화면 이동 구현

- CardForm은 폼 태그로 구현, RegisterComplete는 그냥 버튼으로 구현
- 목적성에 맞게 결정함

- 상태를 확실히 초기화해주기 위해 브라우저 페이지를 새로 로드하도록 함
- 사용자 경험 측면에서도 작업 완료 후 새로고침은 자연스러운 UX라고 판단함

sooyeoniya committed:
feat: 입력 값이 모두 filled 상태일 때 확인 버튼 활성화

sooyeoniya committed:
feat: 입력 값에 오류가 없는 경우 확인 버튼 활성화

- isFormComplete와 areAllFieldsFilled, areAllFieldsValid를 최적화할지 고민을 해봤다
- 하지만 단순 불리언 값을 체크하는 간단한 연산이기도 하고, 자주 값이 변경되긴 하지만, 성능 병목이 생기는 수준은 전혀 아님
- 그리고 이 값들은 단순히 제출 버튼을 활성화할지 비활성화할지를 결정하는 용도로만 사용되고, 다른 복잡한 UI를 업데이트 하지 않음
- 그리고 입력 필드가 총 5개로 계산량이 많지 않다고 판단함

- 이런 상황에서는 코드 가독성과 유지보수성이 성능보다 더 중요한 요소임
-> useCallback, useMemo를 사용하는 것보다 지금 구조를 유지하는 것이 더 직관적이라고 판단함

- 실제로 UI에서 성능 병목 현상이 발생하는 경우에는 최적화가 필요 -> profiler를 통해 직접 확인하기

sooyeoniya committed:
fix: router 관련 storybook 오류 수정

- router 관련 로직은 MemoryRouter로 감싸주어야 함
- 기본 pathname 설정 및 useLocation으로 넘어온 state 초깃값 정의

sooyeoniya committed:
feat: 입력 요소가 채워지면 다음 폼 보이도록 구현

sooyeoniya committed:
chore: useSequentialForm 조금 더 명확한 함수/변수명으로 수정

sooyeoniya committed:
feat: 입력 전부 채워졌을 때 포커스 이동 기능 구현

sooyeoniya committed:
docs: README.md 구현한 기능 체크 표시

sooyeoniya committed:
fix: 404 error 수정

- link 태그 문제: 없는 css 파일 제거 및 svg 파일 대체

sooyeoniya committed:
fix: 배포 시 라우터 Not Found 에러 수정

- GitHub Pages 서버측 라우팅 지원하지 않기 때문에 /react-payments 경로를 인식하지 못함

- basename 경로 추가
- 로컬 개발 환경과 배포 환경 구분하기 위해 환경 변수로 URL 분기 처리

sooyeoniya committed:
fix: 확인 버튼 클릭 시 초기 URL도 수정

- basePath 적용

sooyeoniya committed:
refactor: 선택된 카드 찾는 로직 수정

- 기존에 있는 객체 배열 활용 -> 중복 코드 제거

sooyeoniya committed:
refactor: 라우터 상수 위치 변경

- config 폴더 생성에서 관리하도록 함(환경 설정 관련 상수 폴더)

sooyeoniya committed:
refactor: 상수, 타입, 유틸 파일 정리

- validations 유틸 도메인 별로 나눈 파일 모두 관심사별로 통합

sooyeoniya committed:
refactor: useSequentialForm 객체로 반환

sooyeoniya committed:
refactor: useSequentialForm 중복 로직 통합

sooyeoniya committed:
refactor: CardValidation 에러 상태 업데이트 로직 분리

- 중복 코드 통합

sooyeoniya committed:
chore: errors -> error 단수형으로 통일

sooyeoniya committed:
chore: config -> configs 폴더명 복수형으로 통일

sooyeoniya committed:
chore: cardNumbers -> cardNumber 단수형으로 통일

sooyeoniya committed:
chore: 유효기간 관련 에러 메시지 구체화

sooyeoniya committed:
chore: ComponentProps로 변경

- Select, Input의 경우에는 ref 속성도 포함해야 하므로 ComponentProps로 변경하는 것이 좋음
- 기존 방식은 순수 HTML 속성들만 포함하기 때문에 ref, key, children까지 포함하는 React 특화 타입을 사용하려면 ComponentProps를 사용해야 함

sooyeoniya committed:
docs: README.md 컴포넌트 관계도 정리

- 렌더 트리 형식으로 정리

sooyeoniya committed:
docs: README.md 문서 수정

sooyeoniya committed:
docs: context-api 관련 트러블 슈팅 문서 작성

sooyeoniya committed:
docs: 컴포넌트 분리 관련 트러블 슈팅 문서 정리

sooyeoniya committed:
Merge branch 'woowacourse:main' into step2

JeongBin0227 commented:
[Form 상태 관리 & Custom Hook 분리]

반복되는 로직을 custom hook으로 분리했는가?
- Card 입력 관리, 입력 유효성 검사, 입력 포커싱 제어, Form 흐름 제어 등 주요 로직을 잘 분리하신거 같아요 👍 

[개선점]
대부분 잘하셨는데, 걱정되는건 훅 간의 의존성 명확히 관리하면 더 좋을거같아요 
이게 무슨말이나면, useCard, useCardValidation, useFieldFocus를 보면 서로 의존하는 부분이 있는데, 나중에 규모가 더 커지면 훅 간 순환 참조 위험이 생길 수도 있어요 최대한 각 훅은 최대한 독립성 있게 구성하는 걸 습관처럼 가져가면 더 좋을거 같아요

hook 내부와 UI 컴포넌트의 역할이 명확하게 분리되어 있는가?
- CardNumberInput 을 보면, 훅은 오직 상태 관리와 비즈니스 로직만 담당하고, 컴포넌트는 UI와 이벤트 바인딩에만 집중하고 있는거 같아요! 

상태 흐름이 직관적이며, Form 전체를 일관되게 관리할 수 있는 구조인가?
- CardForm → CardFormFields → 각각 Input들로 흐름이 일관되게 이어져서 가독성도 좋은거 같아요.
특히 useCard, useCardValidation이 Form의 중심 상태를 잘 잡아주고 있는것 같습니당
- 유효성 검증 흐름도 validateCardNumbers, validateExpiration, validateCvc 식으로 통일되어 있어요! 👍

JeongBin0227 commented:
[입력 UI 흐름과 UX]

카드 번호 입력 시 필드 간 자동 포커싱 이동이 자연스럽게 동작하는가?
- 코드상으로 useFieldFocus 이쪽이 담당하는거 같은데, 서비스 상으로 확인해도 잘 이동하는거 같아요!
- 보통 이런식으로 구현은 onChange, onBlur 둘중 하나를 선택해 구현할수 있는데, 니야는 지금 상황에서는 onChange 가 적절했는데 잘 구현하신거 같아요~ (참고로 onBlur 는 4글자를 다 치더라도 Input을 벗어날  까지 기다렸다가 그제야 다음 필드로 이동해서 지금과 맞지 않아요~)


숫자만 입력 가능한 필드에서 제한, 에러 메시지, 유효성 피드백 등 사용자 경험이 충분히 고려되었는가?
- 태그의 벨리데이션도 잘 이용했고, JS 도 잘 활용하고 있는거 같아요 👍

JeongBin0227 commented:
[컴포넌트 구조 및 재사용성]
컴포넌트가 명확한 역할과 책임을 가지며 과도하게 분리되거나 중첩되어 있지 않은가?
- UI 가 너무 과도하게 나눠지지도 않았고, 관심사도 잘 지켜지면서 분리된거 같아요! 👍 

Input, Button 등 재사용 가능한 컴포넌트가 잘 정의되어 있는가?
- 작은 컴포넌트가 재상용이 좋게 잘 정의되어 있네요

[개선점]
좀더 범용적으로 쓰이기 위해 지금 Input은 기본적인 value, onChange만 처리하고 있는데, 다양한 type이나 autoComplete, inputMode 같은 것도 props로 열어주면 더 완성도 높아질 거 같아요!

[참고사항]
이건 이 과제에서는 필요없는 부분이었는데, 알아 두면 좋을거 같아서 적어요~
Input이나 Button이 스타일이 다를 경우, 스타일 variation 지원 고려해보면 좋을거 같아요!
그럼 매번 새 컴포넌트 만들지 않고, variant prop 추가해서 스타일 분기하면 더 좋을거 같아요~

JeongBin0227 commented:
4. 상태 기반 유효성 검사 및 확인 버튼 활성화
모든 필드가 유효할 때만 확인 버튼이 정확히 활성화/비활성화되는가?
-isFormValid 를 공유함으로써 유효할때만 확인 버튼이 잘 동작하게 작업했고, 화면으로 봐도 잘 동작합니다!

유효성 검사의 기준이 명확하고, 상태 변경에 따른 UI 반응이 잘 연결되어 있는가?
- 에러 상태는 즉시 Input에 반영되거나, submit 버튼 상태에 반영됩니다 👍 

[개선점]
- 현재 validateCardNumbers, validateExpiration, validateCvc가 따로 존재하는데 패턴이 비슷하니까 공통 유틸을 만들어 사용하면 더 좋을거 같아요

[참고사항]
- 이번 요구사항은 아니지만, 현재는 "숫자만 입력 가능합니다"처럼 간단한 에러만 보여주고 있는데, 사용자 경험을 위해 좀더 자세한 에러를 보여주면 좋을거 같아요~

JeongBin0227 commented:
> 1️⃣ useEffect 적절성

-> 전체적으로 useEffect 를 확인해봤는데, 불필요하게 사용된 useEffect 나 의존성이 없는거 같아요!

JeongBin0227 commented:
> 2️⃣ useState와 useRef를 적절하게 잘 사용했는가?

-> 이것도 역시 useState 와 useRef 가 필요할때는 적절히 알고 잘 사용하신거 같아요~

JeongBin0227 commented:
> 3️⃣ 성능 최적화 관점에서 문제가 발생할만한 부분이 있는가?

-> 최적화를 진행한다고 무분별하게 사용하는 경우가 많은데, 복잡한 계산이 필요한 컴포넌트가 없었고, 리렌더링 부하가 크지 않은 단순 Input 흐름 위주였기 때문에 크게 필요가 없었다고 생각해요!

JeongBin0227 commented:
> 📌 페이지 단위의 Storybook UI 테스트에 대하여

-> 우선 결론부터만 말씀드리면, 제가 이 과제에 대한 구현사항을 봤을때는, 
지금은 작은 컴포넌트 테스트에 집중하고, 페이지 Story는 지우거나 아주 가볍게만 남기는게 좋을거 같아요. Interaction까지 신경쓰는 건 다음 단계나 방학때 진행해보면 좋을거 같아요
이유는 이번 미션은 컴포넌트에 집중을 하는게 주요 미션이라고 생각해요. 그래서 Form 전체를 다루는 Interaction 보다는 컴포넌트 + 작은 단위의 상태 테스트에 집중하는 게 좀더 맞지 않을까.. 생각했어요 그리고 storybook을 사용하는 이유도 컴포넌트 단위 테스트에 초점을 맞춰 놓은거라 즉, 작은 단위 UI (Button, Input, CardPreview...)를 고립된 환경에서 다양한 상태를 독립적으로 검증하는 게 목적이기 때문에 더욱더 Interaction 는 괜찮다고 생각했어요
오히려 현업에서는 페이지 단위(전체 Form 페이지)는 보통 Storybook보단 Cypress, Playwright 같은 E2E 테스트 도구로 검증합니당

JeongBin0227 reviewed (changes requested):
안녕하세요 니야!

학습목표나, 니야의 질문에 답을 하려면 불가피하게 전반적인 코드를 계속 봐야해서 하루가 넘어버렸네요 ㅠㅠ 죄송합니다 1차로는 학습목표와 니야의 질문에 맞춰서 코멘트를 남겨봤어요~ 이를 먼저 확인해주시고 다시 리뷰 요청을 주시면 그때는 실제 서비스, 스토리북, 코드 단위로 리뷰를 하려고 합니다~ 코드 단위로 볼때 혹시 니야가 더 깊게 봐줬으면 좋겠다는 파일이 있다면 함께 남겨주세요~

미션하느라 고생하셨습니다~

sooyeoniya commented:
> [개선점]
대부분 잘하셨는데, 걱정되는건 훅 간의 의존성 명확히 관리하면 더 좋을거같아요
이게 무슨말이나면, useCard, useCardValidation, useFieldFocus를 보면 서로 의존하는 부분이 있는데, 나중에 규모가 더 커지면 훅 간 순환 참조 위험이 생길 수도 있어요 최대한 각 훅은 최대한 독립성 있게 구성하는 걸 습관처럼 가져가면 더 좋을거 같아요

의존성 관련 피드백 주셔서 감사합니다! 이번 피드백 덕분에 의존성과 순환 참조에 대해 더 깊이 고민해볼 수 있었어요 😃

그런데 한 가지 궁금한 점이 있어 질문드립니다! 🙋🏻‍♀️

`useCard`, `useCardValidation`, `useFieldFocus` 훅들이 서로 의존하고 있다고 말씀해주셨는데, 제가 보기에는 이 3개의 훅 모두 서로 **독립적으로 동작**하는 구조라고 생각했습니다. 혹시 제가 놓치고 있는 부분이 있을까요?

- **useCard / useCardValidation**: 각각 `CardContext`와 `CardValidationContext`에 접근하는 역할만 담당하며, 다른 훅에 직접적으로 의존하고 있지 않고, Context 내부 상태에 대한 비즈니스 로직만 다루고 있습니다.
- **useFieldFocus**: 각 입력 필드에 대한 ref와 포커스 이동을 관리하는 순수한 훅으로, 다른 훅에 의존하지 않고 있습니다. 이 훅을 호출하는 쪽에서 입력 필드의 순서를 담은 ID 배열만 인자로 받아 내부 로직에서 이를 활용하고 있습니다. ref 연결도 호출하는 컴포넌트에서 독립적으로 처리되기 때문에, 이 훅은 의존성 없이 독립적으로 사용할 수 있는 구조를 갖추고 있습니다!

제가 이해한 바로는 순환 참조가 발생하려면, 예를 들어 `useCard`가 `useCardValidation`을 사용하고, 동시에 `useCardValidation`도 다시 `useCard`를 참조해야 하는 상황이어야 하는데요. 현재 구조는 단방향 의존만 존재하고, 외부 훅을 내부에서 직접 호출하거나 참조하지 않기 때문에 순환 참조의 가능성은 낮다고 판단했습니다.

혹시 제가 인지하지 못한 간접적인 의존이나 구조상 문제가 될 수 있는 부분이 있다면 짚어주시면 정말 감사하겠습니다! 🙇🏻‍♀️

sooyeoniya commented:
# 💡의존성 관련 고민

의존성 관련 피드백 덕분에 제 코드에서 의존성이 강해질 수 있는 구조를 한 번 점검해봤는데요.
이 과정에서 **의존성 때문에 잠재적으로 문제가 발생할 수 있는 훅**을 두 개 발견했습니다.

### **📌 useCardSubmit와 useSequentialForm 역할**

이 두 훅의 역할을 간단히 정리해보고, 의존성과 관련된 구조적인 문제도 분석해 보았습니다!

- **useCardSubmit**: 카드 정보가 모두 유효하게 입력 되었는지 확인하고, 폼 제출 시 결과 페이지로 이동시키는 역할을 담당합니다.
- **useSequentialForm**: 카드 등록 폼의 각 입력 섹션을 순차적으로 보여주는 훅인데요, 이 과정에서 현재 섹션이 올바르게 작성되었을 경우에만 다음 섹션을 보여주는 역할을 담당합니다.

이 두 훅은 각 역할 특성상 `useCard`(카드 입력 상태 관리)와 `useCardValidation`(카드 입력 유효성 검사 상태 관리)에 의존할 수밖에 없는 구조입니다.

### **📌** useCardSubmit이 의존하는 정보

- **useCard**
    - **cardCompany:** 카드사 정보 (`/register-complete` 페이지로 전달해야 함)
    - **cardNumber:** 카드 번호 앞 4자리 정보 (`/register-complete` 페이지로 전달해야 함)
    - **areAllFieldsFilled:** 모든 입력 필드가 전부 채워졌는지 여부 (확인 버튼 표시 조건)
- **useCardValidation**
    - **areAllFieldsValid:** 모든 입력값이 전부 유효한지 여부 (확인 버튼 표시 조건)

### **📌** useSequentialForm이 의존하는 정보

- **useCard**
    - **isFieldFilled:** 특정 필드가 모두 입력 되었는지 여부 확인
- **useCardValidation**
    - **hasErrorByType:** 특정 필드에 유효성 오류가 있는지 확인

이 구조는 `useCard`와 `useCardValidation`의 내부 구현에 강하게 결합되어 있기 때문에, 훅 간 상호 의존성이 생길 가능성이 있고, 이 때문에 케빈의 말씀대로 순환 참조가 발생할 위험도 있다고 생각했습니다. 또한, 이 두 개의 훅 모두 외부 훅의 상태와 로직에 의존하고 있어서 재사용성도 낮다는 문제가 있습니다.

그래서 케빈이 앞서 말씀하신대로 **의존성을 줄이고 훅을 최대한 독립적으로 만들기 위해, 재사용성을 높인 범용적인 훅으로 구성하는 것이 좋을지**에 대해 고민해보았습니다.

### **📌** useCardSubmit의 재사용성에 대한 고민

`useCardSubmit`은 꼭 범용적인 훅으로 만들 필요는 없겠다는 생각이 들었어요! 아래처럼 정리해봤는데, 혹시 제가 정리한 방향에 의문점이 있다면 말씀해주시면 감사하겠습니다! 😄

1. **도메인에 맞는 명확한 책임**
    
    이 훅은 말 그대로 `카드 제출`에 특화된 훅이라서, 그 자체로 이미 역할이 명확하고 구체적인 것 같아요. 그래서 굳이 범용적인 훅처럼 추상화하기보다는, 카드 도메인에 최적화된 구조가 더 직관적이고 유지보수에도 유리하지 않을까 싶었습니다.
    
2. **강한 도메인 의존성**
    - `cardCompany`, `cardNumber` 같은 카드 관련 데이터 사용
    - `CARD_COMPANY_INFO`라는 카드사 정보 상수에 의존
    - `/register-complete` 같은 명확한 라우터 경로 사용
    
    이런 요소들 때문에, 재사용 보다는 도메인 안에서만 쓰는 게 맞는 훅 같습니다.
    
3. **비즈니스 로직 특성상 재사용 어려움**
    
    이 훅 내부에서, 카드 번호의 앞 4자리와 카드사 이름을 등록 결과 페이지로 전달하는 구조는 다른 도메인에서는 쉽게 재사용하기 어려운 방식이라고 생각합니다.
    
4. **불필요한 추상화**
    
    이 훅을 제네릭화 하거나 의존성을 외부에서 주입하는 방식도 고민은 해봤는데요. 지금처럼 제출 로직이 하나뿐인 상황에서 그렇게까지 복잡하게 구성하는 건 오히려 가독성만 떨어지고, 오버엔지니어링일 수 있겠다 싶었습니다. 나중에 유사한 제출 흐름을 가진 폼이 여러 개 생긴다면 그땐 분리해도 좋을 것 같다고 판단했습니다!
    
5. **순환 참조 관련 문제**
    
    현재 구조에서는 `useCardSubmit`이 `useCard`랑 `useCardValidation`을 참조하지만, 반대로 참조되는 흐름은 없어서 순환 참조 문제는 크지 않다고 생각합니다.
    
    ```
    useCardSubmit → useCard
    useCardSubmit → useCardValidation
    ```
    
    - `useCard`: 각 필드의 입력 상태를 제공하는 역할 (Provider)
    - `useCardValidation`: 유효성 검사 로직만 담당 (Provider)
    - `useCardSubmit`: 폼 제출 처리만 담당 (Consumer)
    
    각각의 훅이 명확하게 역할이 나뉘어 있어서 서로의 책임을 침범하지 않고 있다고 생각해요.
    
    물론, 추후 구조가 바뀔 경우에는 아래와 같은 부분만 지켜주면 괜찮을 것 같다는 생각이 듭니다!
    
    - `useCard`나 `useCardValidation` 안에서 `useCardSubmit`을 사용하지 않기
    - 새로운 훅 만들 때도 의존성 방향을 일관되게 유지하기

결론적으로, 이 훅은 카드 폼 도메인에 특화되어있는 명확한 목적과 책임을 가진 훅이기 때문에, 굳이 범용적으로 추상화할 필요는 없을 것 같다는 생각이 들었습니다.

과하게 의존성을 분리해서 추상화하기보다는, 오히려 지금처럼 도메인에 특화된 구조가 코드 품질 측면에서 더 직관적이고, 유지보수하기도 쉽다고 생각했어요!

혹시 더 나은 구조나 관점 있으시디면 피드백 주시면 감사하겠습니다! 😊

### **📌** useSequentialForm에서 의존성 줄이기 위한 고민

`useCardSubmit`과는 좀 다르게, `useSequentialForm` 훅은 의존성을 분리하여 재사용성 관점에서 접근해보는 게 더 의미 있을 것 같다고 생각했어요!

물론 지금 미션 기준으로 보면 아직 다른 폼이 존재하는 건 아니라서, 굳이 분리하려다 보면 오히려 복잡하고 이해하기 어려운 구조가 될 수 있을 것 같긴 해요. 약간 오버엔지니어링 느낌도 들고요 😅

그래도 이 훅이 다른 폼들에서도 충분히 활용 가능할 것 같다는 생각이 들어서, 왜 `useCardSubmit` 보다 재사용성 관점에서 더 가치가 있다고 생각했는지, 그 이유와 사용 가능한 부분들에 대해서 정리해보고 어떤 방향성으로 의존성을 분리할 수 있을지에 대해서는 한 번 고민해보고 정리해봤습니다!

우선 현재 해당 훅의 핵심 기능은 다음과 같습니다.

1. 각 단계별 완료 상태 관리
2. 순차적으로 폼 필드를 표시
3. 완료된 필드에 따라서 다음 필드 자동으로 활성화

이 패턴 자체가 재사용 관점에서 보면 사실 카드 폼 뿐만 아니라, 회원가입 폼, 설문조사, 주문 단계 등 여러 UI에서 일반적으로 자주 나오는 패턴이라고 생각했습니다. 따라서, 이 훅의 핵심 로직인 “순차적인 폼 표시 관리”는 특정 도메인(카드 폼)에만 국한되는 것이 아니라고 생각했어요!

현재 이 훅의 의존성은 이런 식으로 카드 폼에 붙어있는 상태입니다.

```tsx
// 카드 폼의 순서를 나타내는 각 필드의 ID 배열
const formSequence = [
  CARD_FORM_TYPE.cardNumber,
  CARD_FORM_TYPE.cardCompany,
  CARD_FORM_TYPE.expirationPeriod,
  CARD_FORM_TYPE.cvcNumber,
  CARD_FORM_TYPE.password,
];

// 초기 상태 정보
const initialState = {
  [CARD_FORM_TYPE.cardNumber]: false,
  [CARD_FORM_TYPE.cardCompany]: false,
  [CARD_FORM_TYPE.expirationPeriod]: false,
  [CARD_FORM_TYPE.cvcNumber]: false,
  [CARD_FORM_TYPE.password]: false,
};

// CardFormType이라는 타입에 직접 의존
const [visibleFields, setVisibleFields] = useState<
    Record<CardFormType, boolean>
  >({ ...initialState, [CARD_FORM_TYPE.cardNumber]: true });
  // 첫 번째 cardNumber에 대한 필드만 true로 활성화 해두기
  
const { hasErrorByType } = useCardValidation();
const { isFieldFilled } = useCard();
```

카드 도메인과 관련된 내부 상태나 훅에 깊이 의존하고 있어서, 지금 상태로는 다른 곳에서 재사용하기가 어렵습니다.

그래서 이런 식으로 개선할 수 있을 것 같다고 생각했습니다.

```tsx
interface UseSequentialFormProps<T extends string> {
  fieldSequence: T[]; // 필드 순서를 담은 배열
  validationFn: (field: T) => boolean; // 특정 필드에 대한 유효성 검증 함수 (기존의 hasErrorByType)
  completionFn: (field: T) => boolean; // 특정 필드에 대한 입력 완료 검증 함수 (기존의 isFieldFilled)
  initialField?: T;  // 선택적으로 받도록 함, 기본값은 첫 번째 필드로 설정하면 됨
}
```

이 부분은 해당 훅이 매개변수로 받는 인터페이스인데요, 의존성 역전 원칙(DIP)을 적용해서 필요한 데이터나 함수를 매개변수로 주입받게 바꿔보았습니다.

이렇게 하면 카드 폼에 대한 직접 의존 없이도 동작하게 되니까, 다른 프로젝트나 도메인에서도 쉽게 가져다 쓸 수 있는 범용 훅이 될 수 있을 것 같다는 생각이 들었어요.

하지만, 아쉬운 점은

- 이 구조로 바꾸면 사용할 때 옵션을 많이 넘겨줘야 해서, 사용하는 입장에서는 오히려 **복잡하고 덜 직관적으로** 느껴질 것 같았습니다.
- 또, 지금처럼 하나의 카드 폼만 있는 상황에선 이렇게까지 추상화하는 건 **현재 요구사항에 비해 과한 설계**일 수도 있겠다 싶었습니다.

결론적으로는, **재사용성 측면에선 괜찮은 시도**가 될 수 있겠지만, 지금 미션 범위에서는 조금 과할 수도 있다는 생각이 들었습니다. 추후에 유사한 패턴이 반복되는 폼이 생긴다면, 그때 구조를 확장하거나 추상화 해도 늦지 않을 것 같아요! 혹시 더 좋은 구조가 있다면 공유 부탁드립니다!! 😁

sooyeoniya committed:
refactor: 페이지 story 문서 상 목록 표시 변경

- 페이지 story 테스트에 대하여 불필요하다고 판단했지만, 이미 작성한 내용이므로 지우지 않도록함
- 대신, 스토리북 문서에서 다른 작은 단위의 컴포넌트와 헷갈릴 우려가 있어서, 추가 경로를 주어 목록을 다르게 표시하도록 함

sooyeoniya commented:
> -> 우선 결론부터만 말씀드리면, 제가 이 과제에 대한 구현사항을 봤을때는,
지금은 작은 컴포넌트 테스트에 집중하고, 페이지 Story는 지우거나 아주 가볍게만 남기는게 좋을거 같아요. Interaction까지 신경쓰는 건 다음 단계나 방학때 진행해보면 좋을거 같아요

이 부분에 대해서 이와 같이 수정해봤습니다! https://github.com/woowacourse/react-payments/pull/472/commits/51ae28fbcb635062e3e23339678d51a62f24973c

페이지 Story를 지우려다가, 이미 구현한 페이지 Story가 조금은 아깝다는 생각이 들어서 일단 그대로 두기로 했습니다!

다만, 스토리북 문서 상에서 봤을 때 다른 작은 단위의 컴포넌트와 헷갈릴 우려가 있어서 목록을 분기 처리 하기 위해 경로를 하나 추가해줬어요!
목록에 분기를 다르게 주었더니 훨씬 구분하기 쉬워진 것 같습니다 :)

<img width="312" alt="image" src="https://github.com/user-attachments/assets/0158a446-c321-40fa-a667-9452c6cdb58c" />

sooyeoniya commented:
> [개선점]
좀더 범용적으로 쓰이기 위해 지금 Input은 기본적인 value, onChange만 처리하고 있는데, 다양한 type이나 autoComplete, inputMode 같은 것도 props로 열어주면 더 완성도 높아질 거 같아요!

현재 Input 컴포넌트에서는 이미 ComponentProps를 통해 HTML input 속성을 상속받고 있습니다!

```tsx
import { ComponentProps } from "react";
import { InputStyles } from "./Input.styled";

interface InputProps extends ComponentProps<"input"> {
  isError: boolean;
}

export default function Input({ isError = false, ref, ...props }: InputProps) {
  return <InputStyles $isError={isError} ref={ref} {...props} />;
}
```

때문에, HTML input 요소의 모든 표준 속성(type, autoComplete, inputMode 등)을 상속받고 있고, {…props}를 통해서 모든 속성이 실제 input 요소에 전달되고 있습니다.

실제로 사용되는 부분에서도 여러 HTML input 속성을 사용하고 있구요!

```tsx
// CardCvcInput.tsx 파일 내부
<Input
  key={CARD_FORM_TYPE.cvcNumber}
  placeholder={PLACEHOLDER.cvcNumber}
  maxLength={MAX_LENGTH.cvcNumber}
  isError={cvcNumberError.hasError}
  value={cvcNumber}
  onChange={(e: ChangeEvent<HTMLInputElement>) =>
    handleCvcNumberChange(e.target.value)
  }
  onBlur={(e: ChangeEvent<HTMLInputElement>) =>
    handleCvcNumberChange(e.target.value)
  }
  autoFocus
/>
```

그렇기 때문에 이 부분에 있어서는 충분히 확장성 있는 Input 컴포넌트라고 생각하여 문제 없을 것 같습니다 😁

---

> [참고사항]
이건 이 과제에서는 필요없는 부분이었는데, 알아 두면 좋을거 같아서 적어요~
Input이나 Button이 스타일이 다를 경우, 스타일 variation 지원 고려해보면 좋을거 같아요!
그럼 매번 새 컴포넌트 만들지 않고, variant prop 추가해서 스타일 분기하면 더 좋을거 같아요~

variation에 대한 고민은 확실히 Input이나 Button 같은 컴포넌트를 재사용하는 데 훨씬 좋은 방향인 것 같아요! 이번 미션에선 당장 필요하진 않아서 반영하진 않았지만, 다음 프로젝트나 컴포넌트 범용성을 높여야 할 상황에선 충분히 고려해볼 수 있을 것 같아요. 좋은 인사이트 감사합니다 😊

sooyeoniya commented:
> [개선점]
현재 validateCardNumbers, validateExpiration, validateCvc가 따로 존재하는데 패턴이 비슷하니까 공통 유틸을 만들어 사용하면 더 좋을거 같아요

이 부분에 대해 한 번 고민 해봤는데요, 현재 구현 자체가 이미 꽤 재사용성을 고려해서 작성된 형태라고 생각했어요!

```tsx
const validateCvcNumber = (value: string) => {
  const errorMessage = isErrorCvcNumber(value);
  setCvcNumberError(updateSimpleErrorState(errorMessage));
};

const validateCardCompany = (value: string) => {
  const errorMessage = isErrorCardCompany(value);
  setCardCompanyError(updateSimpleErrorState(errorMessage));
};

const validatePassword = (value: string) => {
  const errorMessage = isErrorPassword(value);
  setPasswordError(updateSimpleErrorState(errorMessage));
};
```

위처럼 각 필드에 대한 검증 함수가 3~4줄 밖에 되지 않아서, 이미 충분히 직관적이고 가독성도 괜찮은 수준이라고 느꼈습니다!

물론 아래처럼 공통 유틸 함수로 추상화하는 것도 가능하긴 한데요,

```tsx
const validateSingleField = <
  T extends { hasError: boolean; errorMessage: ErrorMessage }
>(
  value: string,
  validator: (value: string) => ErrorMessage,
  setErrorState: React.Dispatch<React.SetStateAction<T>>
) => {
  const errorMessage = validator(value);
  setErrorState(updateSimpleErrorState(errorMessage) as T);
};

const validateCvcNumber = (value: string) => {
  validateSingleField(value, isErrorCvcNumber, setCvcNumberError);
};

const validateCardCompany = (value: string) => {
  validateSingleField(value, isErrorCardCompany, setCardCompanyError);
};

const validatePassword = (value: string) => {
  validateSingleField(value, isErrorPassword, setPasswordError);
};
```

실제 이 추상화 함수를 사용하는 부분에서는 깔끔해보일지 몰라도, 이런 식으로 제네릭을 사용한 추상화는 타입이 복잡해지고, 함수 호출할 때도 인자가 3개나 필요하고, 전체적인 코드 길이 자체도 길어져서 오히려 가독성도 떨어뜨리는 것 같더라구요.

게다가 이미 `updateSimpleErrorState`로 한 번 공통 로직을 분리해둔 상태라, 현재 수준의 공통화만으로도 충분히 선언적이고 깔끔한 구조라고 판단했습니다!

그래서 이번엔 굳이 추상화까지는 도입하지 않는 방향이 더 좋을 것 같습니다! 😃

---

> [참고사항]
이번 요구사항은 아니지만, 현재는 "숫자만 입력 가능합니다"처럼 간단한 에러만 보여주고 있는데, 사용자 경험을 위해 좀더 자세한 에러를 보여주면 좋을거 같아요~

혹시 케빈이 원하는 방향이 “숫자만 입력 가능합니다.” 외에도 추가로 어떤 방식으로 입력해주면 좋겠다 같은 방향성을 제시해주는 에러 메시지를 원하신걸까요!?

그러면 “카드 번호는 4자리의 숫자로만 입력해주세요.” → 이런식으로 명세를 수정해 볼 수도 있겠네요!!

그리고 사용자 경험 측면에서 첨언해보자면, 현재 `ERROR_MESSAGES` 상수를 통해 다양한 에러 상황을 관리하고 있어서, 충분히 사용자 경험 측면에서도 어느 정도 고려가 되어 있다고 생각했어요!

```tsx
export const ERROR_MESSAGES = {
  NUMBER: "숫자만 입력 가능합니다",
  INVALID_MONTH: "유효한 월(MM)을 입력해 주세요. (01~12)",
  INVALID_YEAR: "현재 연도 이상의 유효한 연도(YY)을 입력해 주세요.",
  CHOOSE_CARD_COMPANY: "카드사를 하나 선택해주세요",
  FILL_CARD_NUMBER: "카드 번호를 모두 입력해 주세요",
  FILL_DATE: "유효 기간을 모두 입력해 주세요",
  FILL_CVC: "CVC 번호를 모두 입력해 주세요",
  FILL_PASSWORD: "비밀번호를 모두 입력해 주세요",
} as const;
```

유효성 검증은 숫자가 아닌 경우를 가장 먼저 체크하고 있고, 그 외에도 다른 에러들도 순차적으로 검증해서 표시되도록 되어 있습니다.

만약 다른 에러를 제대로 발견 못하신 부분이라면, 아마 우선순위 상 가장 먼저 걸리는 "숫자만 입력 가능합니다" 메시지만 보이셔서 그렇게 느끼셨을 수도 있을 것 같아요! 숫자인 경우에는 이후 단계에서 다른 에러들도 잘 처리되도록 구성되어 있습니다 😊

sooyeoniya commented:
안녕하세요 케빈! 니야입니다 ☺️
오늘부터 꿀같은 4일 연휴인데, 혹시 케빈도 연휴 동안 본업은 잠시 쉬시나요? ㅎㅎ

케빈께서 남겨주신 리뷰 꼼꼼히 읽어보고, 답변 달아두었습니다! 😁
평일에 다른 일정들이 좀 있어서 리뷰 반영이 조금 늦어진 점 죄송해요 🙇🏻‍♀️

그리고 최근에 준 코치께서 추가해주신 코드래빗이 갑자기 등장(?)해서… PR 내용을 조금 읽기 복잡하게 만들어버린 것 같네요 😅

리뷰 답변 확인 한 번 부탁드릴게요! 이번 미션도 꼼꼼하게 봐주셔서 감사합니다 🤓

JeongBin0227 commented:
> 💡의존성 관련 고민
> useCardSubmit 에 대한 고민

-> 니야가 말한 대로 useCardSubmit은 카드 도메인에 강하게 종속되어 있지만, 일부 유효성 확인 + 제출 흐름은 다른 도메인에서도 재활용 가능성이 있지 않을까 생각해요.
예를 들면

```
// 공통 로직
function useFormSubmit({ isValid, isFilled, onSubmit }) {
  const handleSubmit = () => {
    if (!isValid() || !isFilled()) return;
    onSubmit();
  };
  return { handleSubmit };
}

// 도메인 특화 훅 내부에서 위 공통 훅 사용
function useCardSubmit() {
  const { cardCompany, cardNumbers } = useCard();
  const { isFormValid, areAllFieldsFilled } = useCardValidation();
  const navigate = useNavigate();

  const handleComplete = () =>
    navigate('/register-complete', {
      state: {
        cardCompany,
        cardNumber: cardNumbers.first,
      },
    });

  return useFormSubmit({
    isValid: isFormValid,
    isFilled: areAllFieldsFilled,
    onSubmit: handleComplete,
  });
}
```
이런식으로 두면  useCardSubmit은 도메인 로직만 집중하고,
useFormSubmit은 나중에 다른 폼에도 재활용 할 수 있지 않을까 생각됩니다~  물론 지금 미션 범위에서는 꼭 분리할 필요는 없지만, 이렇게 구조적으로 확장할 수 있어서 남겨둡니당~ 👍

JeongBin0227 commented:
> 💡의존성 관련 고민
> useSequentialForm 에 대한 고민

-> 니야가 말씀하신거에 전적으로 동의해요! 지금 미션이 끝난 시점에서 폼은 하나뿐이라 리팩터링은 오버헤드가 더 크고, 실질적 이득은 작을거라고 생각해요. 하지만 지금처럼 구조적으로 “재사용할 수 있을까?”, “의존성을 외부에서 주입받으면 확장 가능한가?” 등의 고민을 하신 부분이 지금은 추상화를 하지 않아도, 나중에 이렇게 사고 할 수 있는 힘이 길러지고, 비슷한 상황에서 해당 구조를 예시로 함께 고민할 수 있게 되어 성장에 밑거름이 될거라 확신합니다~ 고민에 대한 퀄리티도 좋고 너무 좋은 고민을 하신거 같아요!

JeongBin0227 commented:
> 혹시 케빈이 원하는 방향이 “숫자만 입력 가능합니다.” 외에도 추가로 어떤 방식으로 입력해주면 좋겠다 같은 방향성을 제시해주는 에러 메시지를 원하신걸까요!?
> 
> 그러면 “카드 번호는 4자리의 숫자로만 입력해주세요.” → 이런식으로 명세를 수정해 볼 수도 있겠네요!!
> 
> 그리고 사용자 경험 측면에서 첨언해보자면, 현재 ERROR_MESSAGES 상수를 통해 다양한 에러 상황을 관리하고 있어서, 충분히 사용자 경험 측면에서도 어느 정도 고려가 되어 있다고 생각했어요!
> 
> export const ERROR_MESSAGES = {
>   NUMBER: "숫자만 입력 가능합니다",
>   INVALID_MONTH: "유효한 월(MM)을 입력해 주세요. (01~12)",
>   INVALID_YEAR: "현재 연도 이상의 유효한 연도(YY)을 입력해 주세요.",
>   CHOOSE_CARD_COMPANY: "카드사를 하나 선택해주세요",
>   FILL_CARD_NUMBER: "카드 번호를 모두 입력해 주세요",
>   FILL_DATE: "유효 기간을 모두 입력해 주세요",
>   FILL_CVC: "CVC 번호를 모두 입력해 주세요",
>   FILL_PASSWORD: "비밀번호를 모두 입력해 주세요",
> } as const;
> 유효성 검증은 숫자가 아닌 경우를 가장 먼저 체크하고 있고, 그 외에도 다른 에러들도 순차적으로 검증해서 표시되도록 되어 있습니다.
> 
> 만약 다른 에러를 제대로 발견 못하신 부분이라면, 아마 우선순위 상 가장 먼저 걸리는 "숫자만 입력 가능합니다" 메시지만 보이셔서 그렇게 느끼셨을 수도 있을 것 같아요! 숫자인 경우에는 이후 단계에서 다른 에러들도 잘 처리되도록 구성되어 있습니다 😊

처음 코멘트를 남긴 의도는 "카드 번호는 4자리의 숫자로만 입력해주세요." 이런식으로 자세한 에러메시지가 좋다라는 것을 말씀드리고 싶었는데, 니야가 설명해주신것을 보고 에러가 우선순위 로직 구조기반으로 되어있다는것을 추가로 알게 되었네요~ 감사합니다 💯

JeongBin0227 reviewed (approved):
안녕하세요 니야!

수정해주신 코드 모두 잘 확인하였습니다~ 니야가 해주신 고민 자체가 너무 퀄리티가 높아서 제가 니야의 코드를 보면서 생각하고 고민했던거 이상으로 생각해볼 수 있는 계기가 된거 같아요! 이번에 남겨드린 코멘트는 편하게 읽어보시면 좋을거 같고, 코드레빗이 남긴 코멘트도 한번 읽어봐도 좋을거 같아요! 저도 코멘트를 남기고 싶었는데, 이미 코드레빗이 남겨서 남기지 않은것들도 있어서요!

이번 미션은 여기서 머지할게요! 만나서 반가왔어요! 다음에 또 만나요~
